### PR TITLE
Github Actions Rework

### PR DESCRIPTION
**Pull request by jackrieck**
# Summary

Reworking Github Actions flow to prepare for CI deployments

## Jobs
- Clippy
- Rustfmt
- Anchor build
- Prettier check

## Future

- Add `anchor test` steps to test against the local validator (requires 1password keys)
- Add `anchor deploy` script (requires 1password keys)
- Remove `--skip-lint` from `anchor build`

## Questions

- Anything we should add for the turbo repo integration?
- It takes almost 10min to build our anchor programs, this is pretty slow but not terrible for master builds. In the future we could potentially use self-hosted github runners or wait for github to release their "runner class" features: https://github.com/orgs/community/discussions/26678 - Should we disable this job until then?

## Fixes

- Ran prettier against lib/cli
- Ran fmt against all programs
- Clippy found a Looping error so I fixed that as well
- Added `.prettierignore`

### CODE REVIEW

**Comment by long-banana** [comment]
> Add `anchor test` steps to test against the local validator (requires 1password keys)

Why would we need 1password keys here?

**Comment by long-banana** [comment]
> Anything we should add for the turbo repo integration?
> It takes almost 10min to build our anchor programs, this is pretty slow but not terrible for master builds. In the future we could potentially use self-hosted github runners or wait for github to release their "runner class" features: https://github.com/orgs/community/discussions/26678 - Should we disable this job until then?

One of the big benefits of using turbo repo is that we can use vercel backed build caching to drastically reduce these builds times and it should work both for rust programs and js programs. If you look at turbo repo's documentation, you should find how to set that up within actions. Obviously if we are pushing a change that requires all the programs to build again, we won't save time there. But I'd imagine that would be pretty edge case and we should be fine with just taking that hit on the rare occasions it happens.

**Comment by jackrieck** [comment]
> > Anything we should add for the turbo repo integration?
> > It takes almost 10min to build our anchor programs, this is pretty slow but not terrible for master builds. In the future we could potentially use self-hosted github runners or wait for github to release their "runner class" features: https://github.com/orgs/community/discussions/26678 - Should we disable this job until then?
> 
> One of the big benefits of using turbo repo is that we can use vercel backed build caching to drastically reduce these builds times and it should work both for rust programs and js programs. If you look at turbo repo's documentation, you should find how to set that up within actions. Obviously if we are pushing a change that requires all the programs to build again, we won't save time there. But I'd imagine that would be pretty edge case and we should be fine with just taking that hit on the rare occasions it happens.

Cool Ill check that out

**Comment by jackrieck** [comment]
> > Add `anchor test` steps to test against the local validator (requires 1password keys)
> 
> Why would we need 1password keys here?

To run the tests we upload the IDLs to the validator but the program Ids gotta match whats in the SDKs, thats why im always changing the keys in my PR to ones I have locally. We could use 1password (if thats where we are storing them) and pull them down to the test runner during build time

**Comment by long-banana** [comment]
Why not use the `--bpf-program` flag on the solana-test-validator command?

**Comment by cotemper** [comment]
> > Anything we should add for the turbo repo integration?
> > It takes almost 10min to build our anchor programs, this is pretty slow but not terrible for master builds. In the future we could potentially use self-hosted github runners or wait for github to release their "runner class" features: https://github.com/orgs/community/discussions/26678 - Should we disable this job until then?
> 
> One of the big benefits of using turbo repo is that we can use vercel backed build caching to drastically reduce these builds times and it should work both for rust programs and js programs. If you look at turbo repo's documentation, you should find how to set that up within actions. Obviously if we are pushing a change that requires all the programs to build again, we won't save time there. But I'd imagine that would be pretty edge case and we should be fine with just taking that hit on the rare occasions it happens.

I was planning to update github actions to use turborepo after merging the system-wide tests PR.
https://turborepo.org/docs/ci/github-actions
I'd happy to make another PR to do this if you want.

**Comment by jackrieck** [comment]
> > > Anything we should add for the turbo repo integration?
> > > It takes almost 10min to build our anchor programs, this is pretty slow but not terrible for master builds. In the future we could potentially use self-hosted github runners or wait for github to release their "runner class" features: https://github.com/orgs/community/discussions/26678 - Should we disable this job until then?
> > 
> > 
> > One of the big benefits of using turbo repo is that we can use vercel backed build caching to drastically reduce these builds times and it should work both for rust programs and js programs. If you look at turbo repo's documentation, you should find how to set that up within actions. Obviously if we are pushing a change that requires all the programs to build again, we won't save time there. But I'd imagine that would be pretty edge case and we should be fine with just taking that hit on the rare occasions it happens.
> 
> I was planning to update github actions to use turborepo after merging the system-wide tests PR. https://turborepo.org/docs/ci/github-actions I'd happy to make another PR to do this if you want.

Yeah works for me!

**Comment by long-banana** [comment]
> > Why not use the `--bpf-program` flag on the solana-test-validator command?
> 
> Not sure that would work because the JS libs are pointed to specific programIds

Why would it not work? That's what the flag is for. You specify the programId you want the binary to use and it gets deployed there regardless if you have the right signing key.

Here's an example. I don't have the signing key for the token metadata program, but this is how you get it to deploy to the proper program id. It's also how 
`solana-test-validator --no-bpf-jit -r --mint (solana address -k test-ledger/validator-keypair.json) --bpf-program metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s ../vendor/metaplex-program-library/token-metadata/target/deploy/mpl_token_metadata.so --bpf-program tfre5zcihbycEcr1xo67saNvovdmXhQFXPM2obQqRjP ../target/deploy/redemption.so
`

File: .github/workflows/pull-request.yml [hunk 1]
```diff
@@ -1,134 +1,93 @@
...
 
-defaults:
-  run:
-    working-directory: ./rust
-
 on:
-  pull_request:
   push:
-    branches: [master]
+    branches: '**'
```
Comment (long-banana) [inline .github/workflows/pull-request.yml:5]: Can we flip this to just auto-build on `master` for now. I'd like to not burn through our Actions mins budget until we have the turbo caching in place. Otherwise, that 10 mins of build time is going to start adding up quick. We can ensure the turbo caching PR turns this back on for all branches.

**Review by long-banana — CHANGES_REQUESTED** [review]
My comment on the auto-build is all I have left. Make that change and this is good to go.

**Review by long-banana — APPROVED** [review]
